src/script/view_model/content/AppLockViewModel.ts

```diff
@@ -0,0 +1,290 @@
+    return queryTimeout;
+  }
+  return Math.min(queryTimeout, configTimeout);
+};
+
+const getUnfocusAppLockTimeOut = () => getTimeOut(QUERY_KEY.APPLOCK_UNFOCUS_TIMEOUT, 'APPLOCK_UNFOCUS_TIMEOUT');
+const getScheduledAppLockTimeOut = () => getTimeOut(QUERY_KEY.APPLOCK_SCHEDULED_TIMEOUT, 'APPLOCK_SCHEDULED_TIMEOUT');
+
+const isUnfocusAppLockEnabled = () => getUnfocusAppLockTimeOut() !== null;
+const isScheduledAppLockEnabled = () => getScheduledAppLockTimeOut() !== null;
+
+export const isAppLockEnabled = () => isUnfocusAppLockEnabled() || isScheduledAppLockEnabled();
+
+export class AppLockViewModel {
+  appObserver: MutationObserver;
+  headerText: ko.PureComputed<string>;
+  isLoading: ko.Observable<boolean>;
+  isSetupPasswordAValid: ko.PureComputed<boolean>;
+  isSetupPasswordBValid: ko.PureComputed<boolean>;
+  isVisible: ko.Observable<boolean>;
+  localStorage: Storage;
+  modalObserver: MutationObserver;
+  passwordRegex: RegExp;
+  scheduledTimeOut: number;
+  scheduledTimeOutId: number;
+  setupPasswordA: ko.Observable<string>;
+  setupPasswordB: ko.Observable<string>;
+  state: ko.Observable<APPLOCK_STATE>;
+  storageKey: ko.PureComputed<string>;
+  unfocusTimeOut: number;
+  unfocusTimeOutId: number;
+  unlockError: ko.Observable<string>;
+  wipeError: ko.Observable<string>;
+
+  constructor(private readonly clientRepository: ClientRepository, selfUser: ko.Observable<User>) {
+    this.localStorage = window.localStorage;
+    this.state = ko.observable(APPLOCK_STATE.NONE);
+    this.state.subscribe(() => this.stopObserver(), null, 'beforeChange');
+    this.isVisible = ko.observable(false);
+    this.isLoading = ko.observable(false);
+    this.storageKey = ko.pureComputed(() => `${APP_LOCK_STORAGE}_${selfUser().id}`);
+    ko.applyBindings(this, document.getElementById('applock'));
+
+    this.isVisible.subscribe(isVisible => {
+      (<HTMLDivElement>document.querySelector('#app')).style.setProperty(
+        'filter',
+        isVisible ? 'blur(100px)' : '',
+        'important',
+      );
+    });
+
+    this.unfocusTimeOut = Config.FEATURE.APPLOCK_UNFOCUS_TIMEOUT * 1000;
+    this.unfocusTimeOutId = 0;
+
+    this.scheduledTimeOut = Config.FEATURE.APPLOCK_SCHEDULED_TIMEOUT * 1000;
+    this.scheduledTimeOutId = 0;
+
+    this.headerText = ko.pureComputed(() => {
+      switch (this.state()) {
+        case APPLOCK_STATE.SETUP:
+          return t('modalAppLockSetupTitle');
+        case APPLOCK_STATE.LOCKED:
+          return t('modalAppLockLockedTitle');
+        case APPLOCK_STATE.FORGOT:
+          return t('modalAppLockForgotTitle');
+        case APPLOCK_STATE.WIPE_CONFIRM:
+          return t('modalAppLockWipeConfirmTitle');
+        case APPLOCK_STATE.WIPE_PASSWORD:
+          return t('modalAppLockWipePasswordTitle');
+        default:
+          return '';
+      }
+    });
+    this.passwordRegex = new RegExp(ValidationUtil.getNewPasswordPattern(8));
+    this.setupPasswordA = ko.observable('');
```

Comment:
What's the difference between `setupPasswordA` and `setupPasswordB`? Is `setupPasswordB` the rerun of `setupPasswordA`?

Reply:
This is just the two input fields ("Passphrase" & "Repeat passphrase")